### PR TITLE
[REVIEW] Fix dask_cudf group_split behavior to handle upstream rearrange_by_divisions

### PR DESCRIPTION
**Pull request by rjzamora**
The recently-added `group_split_cudf` dispatch function does not handle `rearrange_by_divisions` correctly.  This PR adds a new test (`test_rearrange_by_divisions`), as well as a fix for the primary bug.  There are actually **two** problems with the current implementation:

1. [**Primary Bug**] The output of `group_split_cudf` should be a dictionary; with each key being a consecutive integer index, and each item being a cudf dataframe.  The current output is a list, which is clearly a bug (my mistake).

2. [Discrepancy between dask and dask_cudf] The original index values should be conserved in the `group_split_cudf` output.  In other words, each dataframe row in the output dictionary should have an index value that could be used to reconstruct the original dataframe.  To clarify the difference, here is an example of `rearrange_by_divisions` behavior with both pandas and cudf-based dask dataframes:

**Original DataFrame**:
```
    x         y
0   6  0.392580
1  19 -0.929185
2  14  0.079832
3  10 -0.159517
4   7  0.022222
5   6 -0.427793
6  18 -0.531817
7  10 -0.117476
8  10  0.222079
9   3 -0.767977
```

**Pandas-based `rearrange_by_divisions` Behavior**:
```
    x         y
9   3 -0.767977
0   6  0.392580
4   7  0.022222
5   6 -0.427793
3  10 -0.159517
7  10 -0.117476
8  10  0.222079
2  14  0.079832
1  19 -0.929185
6  18 -0.531817
```

**cudf-based `rearrange_by_divisions` Behavior** (after dictionary fix):
```
    x         y
0   3 -0.767977
0   6  0.392580
1   7  0.022222
2   6 -0.427793
0  10 -0.159517
1  10 -0.117476
2  10  0.222079
0  14  0.079832
0  19 -0.929185
1  18 -0.531817
```

Notice that the index values are not conserved for the cudf-based DataFrame result.

**Important Note**:  In the current draft of this PR, index conservation is ignored by using `check_index=False`  in `test_rearrange_by_divisions`.  Correct behavior can be accomplished by resetting the index twice (before and after `scatter_by_map`), but this seems inefficient.  Please feel free to comment on this here.

### CODE REVIEW

File: python/dask_cudf/dask_cudf/tests/test_core.py [hunk 1]
```diff
@@ -195,6 +195,29 @@ def test_set_index_w_series():
...
+        gdf1 = dd.from_pandas(cudf.DataFrame.from_pandas(df), npartitions=4)
+        divisions = (0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
+
+        expect = dd.shuffle.rearrange_by_divisions(
+            ddf1, "x", divisions=divisions, shuffle="tasks"
+        )
+        result = dd.shuffle.rearrange_by_divisions(
+            gdf1, "x", divisions=divisions, shuffle="tasks"
+        )
+        dd.assert_eq(expect, result, check_index=False)
```
Comment (mrocklin) [inline python/dask_cudf/dask_cudf/tests/test_core.py]: So the second bug would be tested if we change this to `check_index=True`?

**Comment by mrocklin** [comment]
The change here seems good to me.  

What do we need to do to resolve the second issue?

File: python/dask_cudf/dask_cudf/tests/test_core.py (hunk 1 above)
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/tests/test_core.py]: Yes - I am currently thinking through the "best" way to address this.

**Comment by rjzamora** [comment]
> What do we need to do to resolve the second issue?

I am trying to figure this out.  [The first commit in this PR](https://github.com/rapidsai/cudf/pull/2895/commits/ac76dfbf0e7e0e48a2192998abc7e5580695a0c0) addresses the issue, but that approach feels wrong to me.    I am looking into *lower-level* ways to avoid this pands-cudf discrepancy.

**Comment by mrocklin** [comment]
Yeah, maybe this is something that we kick back to @harrism and team?

**Comment by kkraus14** [comment]
> Yeah, maybe this is something that we kick back to @harrism and team?

libcudf has no concept of index column(s) unless absolutely necessary. I'd guess it's more likely that we're just losing the index column(s) by not passing them through Cython into cudf::table constructors erroneously somewhere.

Before kicking over we should confirm we need libcudf behavior changed.

**Comment by mrocklin** [comment]
Ah, my apologies.

**Comment by rjzamora** [comment]
> Before kicking over we should confirm we need libcudf behavior changed.

Right - I am thinking the change(s) should be in python/cython

**Comment by rjzamora** [comment]
**Update**:  I modified the `scatter_by_map` DataFrame API and the `scatter_to_frames` libcudf wrapper to optionally preserve the index of each row after the call to `cpp_scatter_to_tables`.  I also updated `test_dataframe_scatter_by_map` and `test_rearrange_by_divisions` -- Feel free to suggest alternative solutions.

**Comment by harrism** [comment]
Is this still a 0.10 PR? It is currently an orphan.

**Comment by kkraus14** [comment]
Let's push to 0.11.

File: python/cudf/cudf/_lib/copying.pyx [hunk 2]
```diff
@@ -236,14 +236,15 @@ def copy_range(out_col, in_col, int out_begin, int out_end,
...
-def scatter_to_frames(source, maps):
+def scatter_to_frames(source, maps, index=None):
     """
     Scatters rows to 'n' dataframes according to maps
 
     Parameters
     ----------
     source : Column or list of Columns
     maps : non-null column with values ranging from 0 to n-1 for each row
+    index : Column or None
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/copying.pyx]: Suggest making this a column or a list of columns for the MultiIndex case

**Comment by kkraus14** [comment]
@rjzamora is this ready to be reviewed or still in progress?

**Comment by rjzamora** [comment]
> @rjzamora is this ready to be reviewed or still in progress?
 
Sorry - not quite ready yet.  In order to handle categorical columns, I still need to add a workaround for the round-trip problem in scatter_to_frames

**Comment by rjzamora** [comment]
@kkraus14 - I added the necessary categorical fix to `scatter_to_frames` (and improved testing).  So, this should be good for review (assuming CI passes).

File: python/cudf/cudf/core/dataframe.py [hunk 3]
```diff
@@ -3929,7 +3939,7 @@ def scatter_by_map(self, map_index, map_size=None):
 
             # Append empty dataframes if map_size > len(tables)
             for i in range(map_size - len(tables)):
-                tables.append(self.iloc[[]])
+                tables.append(cudf.DataFrame(columns=self.columns))
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py]: Instead of returning an empty typed Index this will return a RangeIndex now, is that intended?

Comment (rjzamora) [reply python/cudf/cudf/core/dataframe.py]: Good point -  I was missing a check for empty-dataframe index types in `test_dataframe_scatter_by_map`, so I didn't notice this mistake.  I added a simple fix with test coverage.

**Pull request merged**